ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/nonuniquexids/ImportNonUniqueBranchTest.java

```diff
@@ -0,0 +1,116 @@
+    private static PeriodicRecovery periodicRecovery;
+
+    @BeforeClass
+    public static void beforeClass() {
+        RecoveryEnvironmentBean environmentBean = BeanPopulator.getDefaultInstance(RecoveryEnvironmentBean.class);
+        List<String> moduleNames = new ArrayList<>();
+        moduleNames.add("com.arjuna.ats.internal.jta.recovery.arjunacore.SubordinateAtomicActionRecoveryModule");
+        moduleNames.add("com.arjuna.ats.internal.jta.recovery.arjunacore.XARecoveryModule");
+        environmentBean.setRecoveryModuleClassNames(moduleNames);
+        environmentBean.setRecoveryBackoffPeriod(1);
+
+        BeanPopulator.getDefaultInstance(JTAEnvironmentBean.class).setXAResourceRecordWrappingPlugin(new XAResourceRecordWrappingPluginImpl());
+
+        periodicRecovery = new PeriodicRecovery(false, false);
+
+        for (RecoveryModule recoveryModule : periodicRecovery.getModules()) {
+            if (recoveryModule instanceof XARecoveryModule) {
+                xaRecoveryModule = (XARecoveryModule)recoveryModule;
+                break;
+            }
+        }
+    }
+
+    @AfterClass
+    public static void afterClass() {
+        periodicRecovery.shutdown(false);
+    }
+
+    @Test
+    public void testWrapped() throws Exception {
+        test(true);
+        Assert.assertEquals("resource commit error", 0, XAResourceImpl.getErrorCount());
+    }
+
+    @Test
+    public void testNotWrapped() throws Exception {
+        test(false);
+        Assert.assertNotEquals("resource commit should have failed", 0, XAResourceImpl.getErrorCount());
+    }
+
+    public void test(boolean wrap) throws Exception {
+        XAResourceImpl.clearErrorCount();
+        ResourceManager resourceManagerA = new ResourceManager("jndi:/A", wrap);
+        ResourceManager resourceManagerB = new ResourceManager("jndi:/B", wrap);
+
+        xaRecoveryModule.addXAResourceRecoveryHelper(resourceManagerA);
+        xaRecoveryModule.addXAResourceRecoveryHelper(resourceManagerB);
+
+        Xid xid = XATxConverter.getXid(new Uid(), false, 131080);
+        Transaction tx = SubordinationManager.getTransactionImporter().importTransaction(xid, 10000);
+        XATerminator xaTerminator = SubordinationManager.getXATerminator();
+
+        XAResource resource1 = resourceManagerA.getResource("a1");
+        XAResource resource2 = resourceManagerB.getResource("b1");
+
+        tx.enlistResource(resource1);
+        tx.enlistResource(resource2);
+
+        xaTerminator.prepare(xid);
+        // ./ObjectStore/ShadowNoFileLockStore/defaultStore/StateManager/BasicAction/TwoPhaseCoordinator/AtomicAction/SubordinateAtomicAction/JCA/_
+
+        Transaction txA = SubordinationManager.getTransactionImporter().getImportedTransaction(xid);
+        periodicRecovery.doWork(); // will take recoveryBackoffPeriod seconds, be patient
+        Transaction txB = SubordinationManager.getTransactionImporter().getImportedTransaction(xid);
```

Comment:
The two getImportedTransaction calls here serve no useful purpose. They were handy to validate my reasoning when running under the debugger, but unless you're inclined to explicitly assertNotEquals(txA, txB) they may well be removed to avoid confusion.